mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -510,23 +513,80 @@ public Association createAssociation(AssociationKey key, AssociationContext asso
+	 * <li>A list of plain values such as {@code String}s, {@code int}s etc. in case there is exactly one row key column
+	 * which is not part of the association key (in this case we don't need to persist the key name as it can be
+	 * restored from the association key upon loading) or</li>
+	 * <li>A list of {@code DBObject}s with keys/values for all row key columns which are not part of the association
+	 * key</li>
+	 * <li>A {@link DBObject} with a key for each entry in case the given association has exactly one row key column
+	 * which is of type {@code String} (e.g. a hash map) and {@link DocumentStoreProperties#MAP_STORAGE} is not set to
+	 * {@link MapStorageType#AS_LIST}. The map values will either be plain values (in case it's single values) or
+	 * another {@code DBObject}.
 	 * </ul>
 	 */
-	private List<?> getAssociationRows(Association association, AssociationKey key) {
-		List<Object> rows = new ArrayList<Object>();
+	private Object getAssociationRows(Association association, AssociationKey key, AssociationContext associationContext) {
+		boolean organizeByRowKey = organizeByRowKey( association, key, associationContext );
 
-		for ( RowKey rowKey : association.getKeys() ) {
-			rows.add( getAssociationRow( association.get( rowKey ), key ) );
+		// transform map entries such as ( addressType='home', address_id=123) into the more
+		// natural ( { 'home'=123 }
+		if ( organizeByRowKey ) {
+			String rowKeyColumn = organizeByRowKey ? key.getMetadata().getRowKeyIndexColumnNames()[0] : null;
+			DBObject rows = new BasicDBObject();
+
+			for ( RowKey rowKey : association.getKeys() ) {
+				DBObject row = (DBObject) getAssociationRow( association.get( rowKey ), key );
+
+				String rowKeyValue = (String) row.removeField( rowKeyColumn );
+
+				// if there is a single column on the value side left, unwrap it
+				if ( row.keySet().size() == 1 ) {
+					rows.put( rowKeyValue, row.toMap().values().iterator().next() );
+				}
+				else {
+					rows.put( rowKeyValue, row );
+				}
+			}
+
+			return rows;
 		}
+		// non-map rows can be taken as is
+		else {
+			List<Object> rows = new ArrayList<>();
+
+			for ( RowKey rowKey : association.getKeys() ) {
+				rows.add( getAssociationRow( association.get( rowKey ), key ) );
+			}
 
-		return rows;
+			return rows;
+		}
+	}
+
+	/**
+	 * Whether the rows of the given association should be stored in a hash using the single row key column as key or
+	 * not.
+	 */
+	private boolean organizeByRowKey(Association association, AssociationKey key, AssociationContext associationContext) {
+		if ( association.isEmpty() ) {
+			return false;
+		}
+
+		if ( key.getMetadata().getRowKeyIndexColumnNames().length != 1 ) {
+			return false;
+		}
+
+		Object valueOfFirstRow = association.get( association.getKeys().iterator().next() )
+				.get( key.getMetadata().getRowKeyIndexColumnNames()[0] );
+
+		// TODO We may relax this to other single-column key types by conversion through grid type
```

Comment:
Not yet, as we are not sure whether we really want this or not. I'll ask for feedback in the announcement post. I'd implement it once there is demand for it.